### ISSUE TITLE
Full game save / persistence / snapshot mega-issue

### ISSUE THREAD

**Issue by ElectroJr** [space-wizards/space-station-14#34084]
This is a collection of issues/tasks that need to be addressed for full game saves to work properly. Some of these are engine issues, but a decent number of the problems are in content, so I've opened it here.

See also:
- space-wizards/RobustToolbox/issues/2406
- space-wizards/RobustToolbox/issues/4856

Issues:
- [ ] Nullspace entity and multi-map support space-wizards/RobustToolbox/pull/5572
- [ ] NodeNets like power and atmos need to be turned into entities.
- [ ] Several prototypes are marked as non-saveable via `MetaDataComponent.MapSavable`, which includes most mobs. I don't know specifically why they are marked as such, but that needs to be fixed.
- Some system data is not serialized / not stored on entities
  - [ ] Device Nets #26424
  - [ ] Wire layouts?
  - [ ] Game Ticker / Round flow 
    - Might've already been fixed? but it and various rule systems used to have fields referencing specific entities/rules or maps.
- Some components likely store data in non-serialized fields
  - [ ] Hands? #20776
  - [ ] Need to check that mind/actors work well when loading entities.
    - Might largely be fixed by null-space entity auto inclusion #34020
    - See also #20773
    - What happens if an entity with a player attached is saved, deleted, and then loaded again? 
    - What happens if it gets loaded multiple times? 
    - What if the player disconnects and reconnects during the deletion & loading?
- [ ] space-wizards/RobustToolbox/issues/3119
- [ ] How do we handle BUIs? 
  - If we save them we need to ensure all the BUI state is serialisable.
  - If we don't save them, for consistency we should probably first close all buis before saving.
- [ ] Prototype uploads
  - The full-save system should add an extra entry for listing all uploaded prototypes
- [ ] Resource uploads
  - The save file should also reference any uploaded resources, and the saving system should write them to files alongside the save file.
- CTRL+F & fix:
  - [ ] `Timer.Spawn()`
    - Maybe fine for some "out-of-game" stuff like votes or ghost-role raffles?
  - [ ] `// TODO full game save`

And then we need to thoroughly test this somehow. Catching all the odd bits and bobs that aren't saved properly will be difficult. If admins are feeling sadistic they could just save & reload an ongoing round and collect the hate mail. Though we should probably start by saving a real round and trying to load it locally.

**Comment by deltanedas** [comment]
> mobs

i think mostly its so you dont map mobs and instead use the spawner, having a tested do not map category would help with this
but when i tested it out, saving a urist wrote all its bodyparts etc in the yml even if they arent damaged or anything, could just be a mapinit thing

**Comment by ElectroJr** [comment]
> game will throw if its not at least NetSerializable, i dont think any states arent Serializable too

I've edited the point to clarify: I meant yaml-serializable, not net-serializable. We could try use net-serialization, but that would generally prevent loading an older save with a newer version, for the same reason the replay client version should to be the same as the actual game client version for loading a replay (or at least, needs the same registered net serializable C# types).

**Comment by pcaessayrs** [comment]
I'd love to start working on this as it opens up a lot of interesting potential for future innovative gameplay (e.g. "FTL in ss14").

**Comment by Roudenn** [comment]
It seems like `EntityCoordinates` can't be saved in YAML right now. `SerializationManager` is very hard to read so i couldn't really find a solution about how to make them YAML savable, so this has to be somehow fixed on the engine first before we can save things like NPCs

**Comment by Roudenn** [comment]
I've made some progress on my local branch, now i can save mobs without minds/HTN and post-init stations to YAML.

After loading this saved post-init map it still has a lot of bugs and errors, but it's still a good step forward.

I still have to make some important engine changes, such as:
- Make more things serializable (or just YAML-serializable)
- Add `TrySaveGame` and `TryLoadGame` to `MapLoaderSystem` that saves all maps and loads them back in just 1 file
- Write tests on RT side (the hardest part tbh because it actually has to test literally ALL interactions before and after loading the game save

And then these content changes are required:
- Edit Mind, NPC, StationRecords systems to better support YAML-saving (station records are already done)
- Fix most common errors that occur after reloading saved post-init map
- Write tests for as much content interactions as possible, before and after game save reloading

After this it shouldn't be too hard to implement full game saves, but for now without uploaded assets & prototypes support, since it will require a new level of technology in the engine.

Probably we'll have to also store a .zip archive with YAML for all entities and a separate folder for uploaded data in Content.Server folder.

Also, some interactions may be broken because of CVars, since some systems, for example FTL travel, depend on them strongly. Maybe we should also store some CVars in a TOML file to make sure everything works fine?

There's still a lot of work, but as far as i understand it now, it's totally possible to implement full game saves without too much effort!

**Comment by Roudenn** [comment]
[test2.yml](https://github.com/user-attachments/files/21957263/test2.yml)

With some engine and content modifications, i was finally able to save an actual post-init map. In this case it's Bagel station with force-preseted Nukeops gamemode. My computer was saving it for like 20 seconds, and it weights almost 16 megabytes, so before we can make full game saves we have to make a lot of optimizations to serializers.

I'll start uploading all my changes soon

**Comment by ingles98** [comment]
Damn i gotta say this has progressed very nicely since I dropped my first PR here 3 years back around the same subject

**Comment by Roudenn** [comment]
There's still a lot of stuff to do, since even after full persistence saves are implemented, they WILL get broken constantly because of unintended interactions that can corrupt the save and make it almost impossible to load
